File: java/src/HowSum.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class HowSum {

    private static final int DEFAULT_CHECK_INT = -999;

    private static Map<Integer, List<Integer>> resultMap = new HashMap<>();

    public static void main(String[] args) {
        System.out.println("This is my how sum program..");
        int sum = 7;

        List<Integer> arrayList = new ArrayList<>();

        howSum(sum, new int[] { 5, 3, 4, 7 }, DEFAULT_CHECK_INT, arrayList);

        System.out.println("Result for sum [" + sum + "] -> " + arrayList);
    }

    private static List<Integer> howSum(int sum, int[] args, int checkInt, List<Integer> arrayList) {

        if (resultMap.containsKey(sum))
            return resultMap.get(sum);

        if (sum == 0 && checkInt != DEFAULT_CHECK_INT) {
            arrayList.add(checkInt);
            return arrayList;
        }
        if (sum < 0)
            return arrayList;

        for (int key : args) {
            int diff = sum - key;

            List<Integer> result = howSum(diff, args, key, arrayList);
            if (result.size() > arrayList.size()) {
                resultMap.put(sum, result);
                return result;
            }
        }

        resultMap.put(sum, arrayList);
        return arrayList;
    }
}

File: python/howsum.py

import sys
sys.setrecursionlimit(10000)


def howSum(sum, args, myKey, myList):

    if sum in data.keys():
        return data[sum]

    if sum == 0 and myKey != checkInt:
        myList.append(myKey)
        return myList

    if sum < 0:
        return myList

    for key in args:
        diff = sum - key

        myTempList = howSum(diff, args, key, myList)

        if len(myTempList) > len(myList):
            data[sum] = myTempList
            return myTempList

    data[sum] = myList
    return myList


print("This is my how sum program to learning Dynamic Programming....")
sum = 7
checkInt = -999
myList = []

data = dict()
howSum(sum, [5, 3, 4, 7], checkInt, myList)
print("Result for sum [" + str(sum) + "] -> " + str(myList))
